File: iosApp/Well/SwiftUI/Theme/Typography.swift
//
// Created by Philip Dukhov on 12/29/20.
// Copyright (c) 2020 Well. All rights reserved.
//

import SwiftUI

struct TextStyle: Equatable {
    let fontWeight: Font.Weight
    let fontSize: CGFloat

    static let h4 = Self(fontWeight: .bold, fontSize: 30)
    static let subtitle1 = Self(fontWeight: .regular, fontSize: 21)
    static let subtitle2 = Self(fontWeight: .bold, fontSize: 17)
    static let body1 = Self(fontWeight: .regular, fontSize: 18)
    static let body1Light = Self(fontWeight: .regular, fontSize: 18)
    static let body2 = Self(fontWeight: .regular, fontSize: 16)
    static let body2Light = Self(fontWeight: .light, fontSize: 16)
    static let caption = Self(fontWeight: .regular, fontSize: 13)
    static let captionLight = Self(fontWeight: .light, fontSize: 13)

    var uiFont: UIFont {
        .systemFont(ofSize: fontSize, weight: fontWeight.toUIFontWeight())
    }

    var font: Font {
        .system(size: fontSize, weight: fontWeight)
    }
}

extension Text {
    @inline(__always) func style(
        _ style: TextStyle
    ) -> Text {
        font(style.font)
    }
}

extension View {
    @inline(__always) func style(
        _ style: TextStyle
    ) -> some View {
        font(style.font)
    }
}

private extension Font.Weight {
    func toUIFontWeight() -> UIFont.Weight {
        switch self {
        case .ultraLight: return .ultraLight
        case .thin: return .thin
        case .light: return .light
        case .regular: return .regular
        case .medium: return .medium
        case .semibold: return .semibold
        case .bold: return .bold
        case .heavy: return .heavy
        case .black: return .black
        default:
            fatalError("\(self) weight not provided")
        }
    }
}

File: androidApp/src/main/java/com/well/androidApp/ui/composableScreens/theme/Typography.kt
package com.well.androidApp.ui.composableScreens.theme

import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.sp

val Typography = androidx.compose.material.Typography(
    h4 = TextStyle(
        fontWeight = FontWeight.Bold,
        fontSize = 30.sp
    ),
    h6 = TextStyle(
        fontWeight = FontWeight.Normal,
        fontSize = 21.sp
    ),
    subtitle1 = TextStyle(
        fontWeight = FontWeight.Bold,
        fontSize = 17.sp
    ),
    subtitle2 = TextStyle(
        fontWeight = FontWeight.W500,
        fontSize = 14.sp
    ),
    body1 = TextStyle(
        fontWeight = FontWeight.Normal,
        fontSize = 18.sp
    ),
    body2 = TextStyle(
        fontSize = 14.sp
    ),
    button = TextStyle(
        fontWeight = FontWeight.W500,
        fontSize = 14.sp
    ),
    caption = TextStyle(
        fontWeight = FontWeight.Normal,
        fontSize = 12.sp
    ),
    overline = TextStyle(
        fontWeight = FontWeight.W500,
        fontSize = 12.sp
    )
)
